server/src/main/java/io/crate/analyze/CreateTableAsAnalyzer.java

```diff
@@ -0,0 +1,137 @@
+                                 InsertAnalyzer insertAnalyzer,
+                                 RelationAnalyzer relationAnalyzer,
+                                 Schemas schemas) {
+        this.createTableStatementAnalyzer = createTableStatementAnalyzer;
+        this.insertAnalyzer = insertAnalyzer;
+        this.relationAnalyzer = relationAnalyzer;
+        this.schemas = schemas;
+    }
+
+    public AnalyzedCreateTableAs analyze(CreateTableAs createTableAs,
+                                         ParamTypeHints paramTypeHints,
+                                         CoordinatorTxnCtx txnCtx) {
+        RelationName relationName = RelationName.of(
+            createTableAs.name().getName(), txnCtx.sessionContext().searchPath().currentSchema()
+        );
+        relationName.ensureValidForRelationCreation();
+
+        List<TableElement<Expression>> tableElements = getTableElementsFromSourceQuery(
+            createTableAs.query(), txnCtx
+        );
+        CreateTable<Expression> createTable = new CreateTable<Expression>(
+            createTableAs.name(),
+            tableElements,
+            Optional.empty(),
+            Optional.empty(),
+            GenericProperties.empty(),
+            false
+        );
+
+        //postponing the analysis of the insert statement, since the table has not been created yet.
+        Supplier<AnalyzedInsertStatement> postponedInsertAnalysis = () -> {
+                Insert<Expression> insert = new Insert<>(
+                    createTableAs.name(),
+                    createTableAs.query(),
+                    Collections.emptyList(),
+                    Collections.emptyList(),
+                    Insert.DuplicateKeyContext.none()
+                );
+                return insertAnalyzer.analyze(insert, paramTypeHints, txnCtx);
+            };
+
+        return new AnalyzedCreateTableAs(
+            createTableStatementAnalyzer.analyze(createTable, paramTypeHints, txnCtx),
+            postponedInsertAnalysis
+        );
+    }
+
+    private DocTableInfo resolveTableInfo(Table table,
+                             CoordinatorTxnCtx txnCtx) {
+        return (DocTableInfo) schemas.resolveTableInfo(
+            table.getName(),
+            Operation.INSERT,
+            txnCtx.sessionContext().sessionUser(),
+            txnCtx.sessionContext().searchPath()
+        );
+    }
+
+    private List<TableElement<Expression>> getTableElementsFromSourceQuery (Query sourceQuery,
+                                                                            CoordinatorTxnCtx txnCtx) {
+        var querySpec = (QuerySpecification) sourceQuery.getQueryBody();
+        List<Relation> relations = querySpec.getFrom();
+        List<SelectItem> selectItems = querySpec.getSelect().getSelectItems();
+
+        List<TableElement<Expression>> allTableElements = relations.stream()
+            .map(relation -> (Table)relation)
```

Comment:
I think this logic here may break if the query isn't a simple `select ... from tbl` because it assumes there is a `QuerySpecification` and that the relations are `Table` instances.


I think to support all kinds of query it would be possible to use the `RelationAnalyzer` to analyze the query, and then to create the required table elements from the `outputs` of the `AnalyzedRelation`.

Reply:
thanks for reviewing and suggestions! will move on